File: rust/2024/01/1/src/main.rs
use std::{
    fs::File,
    io::{BufRead, BufReader, Read},
    time::Instant,
};

fn main() {
    let start = Instant::now();

    let file = File::open("input.txt").expect("opening input file");
    println!("{}", process(file));

    println!("{:?}", start.elapsed());
}

fn process(file: impl Read) -> u32 {
    let mut locations_1 = Vec::new();
    let mut locations_2 = Vec::new();

    for line in BufReader::new(file).lines().map_while(Result::ok) {
        let mut locations = line.split_whitespace();

        let location_1: u32 = locations.next().unwrap().parse().unwrap();
        locations_1.push(location_1);

        let location_2: u32 = locations.next().unwrap().parse().unwrap();
        locations_2.push(location_2);
    }

    locations_1.sort();
    locations_2.sort();

    locations_1
        .into_iter()
        .zip(locations_2)
        .map(|(location_1, location_2)| location_1.abs_diff(location_2))
        .sum()
}

#[cfg(test)]
mod tests {
    use crate::process;
    use stringreader::StringReader;

    #[test]
    fn test_example_result() {
        let lines = StringReader::new(
            "3   4
4   3
2   5
1   3
3   9
3   3",
        );

        assert_eq!(process(lines), 11);
    }
}

File: zig/2024/01/1/main.zig
const std = @import("std");
const file: []const u8 = @embedFile("input.txt");

pub fn main() !void {
    var main_timer = try std.time.Timer.start();
    defer {
        std.debug.print("{}\n", .{std.fmt.fmtDuration(main_timer.read())});
    }

    const result = try process(file);

    std.debug.print("{}\n", .{result});
}

fn process(input: []const u8) !u32 {
    var arena = std.heap.ArenaAllocator.init(std.heap.page_allocator);
    defer arena.deinit();
    const allocator = arena.allocator();

    var locations_1 = std.ArrayList(i32).init(allocator);
    var locations_2 = std.ArrayList(i32).init(allocator);

    var lines = std.mem.tokenizeScalar(u8, input, '\n');
    while (lines.next()) |line| {
        var locations = std.mem.tokenizeScalar(u8, line, ' ');

        const location_1 = try std.fmt.parseInt(i32, locations.next().?, 10);
        try locations_1.append(location_1);

        const location_2 = try std.fmt.parseInt(i32, locations.next().?, 10);
        try locations_2.append(location_2);
    }

    std.mem.sort(i32, locations_1.items, {}, std.sort.asc(i32));
    std.mem.sort(i32, locations_2.items, {}, std.sort.asc(i32));

    var result: u32 = 0;
    for (locations_1.items, locations_2.items) |location_1, location_2| {
        result += @abs(location_2 - location_1);
    }

    return result;
}

test "process with example" {
    const result = process(
        \\3   4
        \\4   3
        \\2   5
        \\1   3
        \\3   9
        \\3   3
    );

    try std.testing.expectEqual(11, result);
}
